src/DevTools/__test__/MockRelayRenderer.test.tsx

```diff
@@ -0,0 +1,193 @@
+import { MockRelayRenderer_artist } from "__generated__/MockRelayRenderer_artist.graphql"
+import { MockRelayRenderer_artwork } from "__generated__/MockRelayRenderer_artwork.graphql"
+import { MockRelayRenderer_artworkMetadata } from "__generated__/MockRelayRenderer_artworkMetadata.graphql"
+import { MockRelayRendererArtistQuery } from "__generated__/MockRelayRendererArtistQuery.graphql"
+import { renderWithLoadProgress } from "Artsy/Relay/renderWithLoadProgress"
+import { ContextConsumer } from "Artsy/SystemContext"
+import cheerio from "cheerio"
+import { mount, render } from "enzyme"
+import * as React from "react"
+import { createFragmentContainer, graphql, QueryRenderer } from "react-relay"
+import { MockRelayRenderer, renderRelayTree } from "../MockRelayRenderer"
+
+jest.unmock("react-relay")
+
+const Metadata = createFragmentContainer(
+  (props: { artworkMetadata: MockRelayRenderer_artworkMetadata }) => (
+    <div>{props.artworkMetadata.title}</div>
+  ),
+  graphql`
+    fragment MockRelayRenderer_artworkMetadata on Artwork {
+      title
+    }
+  `
+)
+
+const Artwork = createFragmentContainer(
+  (props: { artwork: MockRelayRenderer_artwork }) => (
+    <div>
+      <img src={props.artwork.image.url} />
+      <Metadata artworkMetadata={props.artwork} />
+      {props.artwork.artist && (
+        <ArtistQueryRenderer id={props.artwork.artist.id} />
+      )}
+    </div>
+  ),
+  graphql`
+    fragment MockRelayRenderer_artwork on Artwork {
+      image {
+        url
+      }
+      artist {
+        id
+      }
+      ...MockRelayRenderer_artworkMetadata
+    }
+  `
+)
+
+const Artist = createFragmentContainer(
+  (props: { artist: MockRelayRenderer_artist }) => (
+    <div>{props.artist.name}</div>
+  ),
+  graphql`
+    fragment MockRelayRenderer_artist on Artist {
+      name
+    }
+  `
+)
+
+const ArtistQueryRenderer = (props: { id: string }) => (
+  <ContextConsumer>
+    {({ relayEnvironment }) => {
+      return (
+        <QueryRenderer<MockRelayRendererArtistQuery>
+          environment={relayEnvironment}
+          variables={props}
+          query={graphql`
+            query MockRelayRendererArtistQuery($id: String!) {
+              artist(id: $id) {
+                ...MockRelayRenderer_artist
+              }
+            }
+          `}
+          render={renderWithLoadProgress(Artist)}
+        />
+      )
+    }}
+  </ContextConsumer>
+)
+
+const query = graphql`
+  query MockRelayRendererQuery {
+    artwork(id: "mona-lisa") {
+      ...MockRelayRenderer_artwork
+    }
+  }
+`
+
+function renderToString(element: React.ReactElement<any>) {
```

Comment:
Thoughts on `React.ReactElement<any>` vs `JSX.Element` in common code? The former opens the doors to more safety later on and so seems better to use but was wondering your thoughts.

Reply:
Ah, a `ReactNode` is anything that React knows how to render, including strings, which doesn’t make sense for our case. Ok, going with `JSX.Element` 👍